sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -254,24 +255,104 @@ object DecorrelateInnerQuery extends PredicateHelper {
   /**
    * Rewrites a domain join cond so that it can be pushed to the right side of a
    * union/intersect/except operator.
+   *
+   * Example: Take a query like:
+   * select * from t0 join lateral (
+   *   select a from t1 where b < t0.x
+   *   union all
+   *   select b from t2 where c < t0.y)
+   *
+   * We have outer table with attributes [x#1, y#2] and after introducing DomainJoins the subquery
+   * is a Union where the left side has output [a#3, x#4, y#5] with DomainJoin [x#4, y#5]
```

Comment:
I assume x#4, y#5 are from t1? similar below x#7, y#8 are from t2?

if yes why is there a domainjoin [x#4, y#5] for t1? I thought only the references to the outer are needed (i.e. I expected to see just x#4 for t1). Similar for t2.

Is this because of the set op (like union all in this case) and PullupCorrelatedPredicates? I cannot immediately see where the domainjoin uses both legs of the union.

Reply:
All the x and y attributes are from t0. They just are different instances of the attributes - they come from separate joins of the outer table - and have different expression IDs. Will update the comment to explain that more.